core/src/main/scala/com/metamx/tranquility/tranquilizer/Tranquilizer.scala

```diff
@@ -0,0 +1,302 @@
+    *
+    * NB: This can actually be wrong: if *any* message in a batch of messages is dropped, then all messages in that
+    * batch will be marked as dropped (even if they actually made it out). This is a limitation of how Tranquility
+    * tracks which messages were and were not dropped. This could be improved once
+    * https://github.com/druid-io/tranquility/issues/56 is addressed.
+    */
+  def send(message: MessageType): Future[Unit] = {
+    requireStarted()
+    if (log.isTraceEnabled) {
+      log.trace(s"Sending message: $message")
+    }
+    val holder = MessageHolder(message, Promise())
+
+    val (exBuffer, future) = lock.synchronized {
+      while (buffer.size >= maxBatchSize - 1 && pendingBatches >= maxPendingBatches) {
+        if (log.isDebugEnabled) {
+          log.debug(
+            s"Buffer size[${buffer.size}] >= maxBatchSize[$maxBatchSize] - 1 and " +
+              s"pendingBatches[$pendingBatches] >= maxPendingBatches[$maxPendingBatches], waiting..."
+          )
+        }
+        lock.wait()
+      }
+
+      if (buffer.isEmpty) {
+        bufferStartMillis = System.currentTimeMillis()
+        lock.notifyAll()
+      }
+
+      buffer += holder
+
+      val _exBuffer: Option[Buffer[MessageHolder]] = if (buffer.size == maxBatchSize) {
+        Some(swap())
+      } else if (lingerMillis == 0 && pendingBatches < maxPendingBatches) {
+        Some(swap())
+      } else {
+        None
+      }
+
+      (_exBuffer, holder.future)
+    }
+
+    exBuffer foreach sendBuffer
+
+    future
+  }
+
+  /**
+    * Create a SimpleTranquilizerAdapter based on this Tranquilizer. You can create multiple adapters based on the
+    * same Tranquilizer, and even use them simultaneously.
+    *
+    * @param reportDropsAsExceptions true if the adapter should report message drops as exceptions. Otherwise, drops
+    *                                are ignored (but will still be reflected in the counters).
+    * @return a simple adapter
+    */
+  def simple(reportDropsAsExceptions: Boolean = false): SimpleTranquilizerAdapter[MessageType] = {
+    SimpleTranquilizerAdapter.wrap(this, reportDropsAsExceptions)
+  }
+
+  /**
+    * Stop the tranquilizer, and close the underlying Beam. Blocks until closed.
+    */
+  @LifecycleStop
+  def stop(): Unit = {
+    started = false
+    sendThread.interrupt()
+    sendThread.join()
+    Await.result(beam.close())
+  }
+
+  override def toString = s"Tranquilizer($beam)"
+
+  private def requireStarted() {
+    if (!started) {
+      throw new IllegalStateException("Not started")
+    }
+  }
+
+  // Swap out the current buffer immediately, returning it and incrementing pendingBatches.
+  // Must be called while holding the "lock".
+  // Preconditions: buffer must be non-empty and pendingBatches must be lower than maxPendingBatches.
+  private def swap(): Buffer[MessageHolder] = {
+    assert(buffer.nonEmpty && buffer.size <= maxBatchSize)
```

Comment:
is there any harm in wrapping this method in lock.synchronized for good measure? then the only requirement when calling swap would be to have the preconditions set.